BJS2-40651 Project Service : Vacancies - Get Vacancy Feature

drewbal opened the pull request:
5/5 Vacancy Feature

drewbal committed:
Initial commit

drewbal committed:
BJS2-40651 Feature: Create Vacancy - adds VacancyDto

DTO: VacancyDto added

drewbal committed:
BJS2-40651 Feature: Create Vacancy - adds Vacancy Mapper

DTO: VacancyDto optimized
Mapper: VacancyMapper added

drewbal committed:
BJS2-40651 Feature: Create Vacancy - adds controller and create method

VacancyController: createVacancy method added

drewbal committed:
BJS2-40651 Feature: Create Vacancy - create full vacancy from dto

VacancyService: toEntityFromDto method added (add fields ignored by mapper)
ProjectService: getProjectById metod added

drewbal committed:
BJS2-40651 Feature: Create Vacancy - validate data to create the vacancy

VacancyValidator: checks the project is valid; role of the creator is valid;
TeamMemberService: returns the team member by id
DataValidationException: new Exception

drewbal committed:
BJS2-40651 Feature: Create Vacancy - creates new vacancy in system

VacancyService: creates a new vacancy;

drewbal committed:
BJS2-40651 Feature: Create Vacancy - sets status at creation

VacancyDto: VacationStatus field added;
VacancyService: sets OPEN status at creation;

drewbal committed:
BJS2-40651 Feature: Create Vacancy - review fixes

VacancyService: toEntityFromDto is private;
VacancyController: path fixed;
Other: blank lines removed

drewbal committed:
BJS2-40651 Feature: Update Vacancy - controller

VacancyController: new method to update the existing vacancy status

drewbal committed:
BJS2-40651 Feature: Update Vacancy - dto update

VacancyDto: upgradeBy field added
Other: VacancyMapper class moved to other package

drewbal committed:
BJS2-40651 Feature: Update Vacancy - validator update

VacancyValidator: checks the role od updater and the number of candidates for vacancy closure
Exception: InsufficientCandidatesException added

drewbal committed:
BJS2-40651 Feature: Update Vacancy - can update vacancy status to CLOSED

VacancyService: updateVacancyStatus method that changes the status of the vacancy based on candidates amount criteria

drewbal committed:
Merge branch 'Dragon-master-stream7' into BJS2-40651-update-vacancy

drewbal committed:
BJS2-40651 Feature: Update Vacancy - review fixes

VacancyController: api v1 added to path
VacancyService: @Transactional annotation added

drewbal committed:
BJS2-40651 Feature: Delete Vacancy - controller upgrade

VacancyController: delete vacancy method added

drewbal committed:
BJS2-40651 Feature: Delete Vacancy - service upgrade

VacancyService: delete vacancy method added
CandidateService: find candidate by id added

drewbal committed:
BJS2-40651 Feature: Create Vacancy - review fixes

VacancyValidator: logic changed, VacancyRepository injected
VacancyService: validation fix
ProjectValidator: basic validation added

drewbal committed:
Merge branch 'BJS2-40651-create-vacancy' into BJS2-40651-update-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/service/VacancyService.java
#	src/main/java/faang/school/projectservice/validator/VacancyValidator.java
#	src/test/java/faang/school/projectservice/validator/VacancyValidatorTest.java

drewbal committed:
BJS2-40651 Feature: Update Vacancy - review fixes

Other: minor conflict fixes

drewbal committed:
Merge branch 'BJS2-40651-update-vacancy' into BJS2-40651-delete-vacancy

drewbal committed:
BJS2-40651 Feature: Delete Vacancy - deletes vacancy by id

VacancyService: deleteVacancy method added

drewbal committed:
BJS2-40651 Feature: Filter Vacancies - controller upgrade

VacancyController: Filter vacancies method added
FilterVacancyDto: added with 3 filters

drewbal committed:
BJS2-40651 Feature: Filter Vacancies - adds filters

VacancyFilters: Vacancy Title, Salary and WorkSchedule filters added to the application.

drewbal committed:
BJS2-40651 Feature: Filter Vacancies - service upgrade

VacancyService: FilterVacancies method with expansion ability through list of new filters

drewbal committed:
Merge branch 'Dragon-master-stream7' into BJS2-40651-create-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/exception/DataValidationException.java
#	src/main/java/faang/school/projectservice/service/TeamMemberService.java

drewbal committed:
BJS2-40651 Feature: Create Vacancy - merge and conflicts solving

Minor fixes

drewbal committed:
Merge branch 'BJS2-40651-create-vacancy' into BJS2-40651-update-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/mapper/VacancyMapper.java
#	src/main/java/faang/school/projectservice/validator/VacancyValidator.java
#	src/test/java/faang/school/projectservice/controller/VacancyControllerTest.java
#	src/test/java/faang/school/projectservice/validator/VacancyValidatorTest.java

drewbal committed:
BJS2-40651 Feature: Update Vacancy - merge and conflicts solving

Minor fixes

drewbal committed:
Merge branch 'BJS2-40651-update-vacancy' into BJS2-40651-delete-vacancy

drewbal committed:
Merge branch 'BJS2-40651-delete-vacancy' into BJS2-40651-filter-vacancy

drewbal committed:
BJS2-40651 Feature: Get Vacancy by id

VacancyController: method to get vacancy by id added
VacancyService: new method for getting vacancy dto by id added

denis-levkhov commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,96 @@
+        Stream<Vacancy> vacancies = vacancyRepository.findAll().stream();
+        return vacancyFilters.stream()
+                .filter(filter -> filter.isApplicable(filters))
+                .flatMap(filter -> filter.apply(vacancies, filters))
```
Это круто!

denis-levkhov commented on src/test/java/faang/school/projectservice/filter/vacancy/VacancyTitleFilterTest.java:
```diff
@@ -0,0 +1,69 @@
+import java.util.List;
+import java.util.stream.Stream;
+
+import static org.junit.jupiter.api.Assertions.*;
```
Импорт всех асершенов через * - круто, возьму на вооружение

denis-levkhov reviewed (commented):
Даже добавить нечего, шикарный профессиональный код.
С удовольствием просмотрел и сделал для себя заметки.
Очень много крутых фишек.
Крутая работа Андрей! (:

denis-levkhov commented on src/main/java/faang/school/projectservice/validator/VacancyValidator.java:
```diff
@@ -0,0 +1,50 @@
+
+    public void validateVacancyCreatorRole(VacancyDto dto) {
+        TeamMember teamMember = teamMemberService.getTeamMemberByUserId(dto.getCreatedBy());
+        if(teamMember.getRoles().stream().noneMatch((ROLES_TO_CREATE_VACANCY::contains))) {
```
Пробел после If

RuslanZinovyev commented on src/main/java/faang/school/projectservice/controller/VacancyController.java:
```diff
@@ -0,0 +1,46 @@
+@RestController
+@RequestMapping("/api/v1/vacancies")
+@RequiredArgsConstructor
+public class VacancyController {
```
В целом контроллер выглядит хорошо, но если добавишь логгер будет вообще по красоте. Можешь потом потестировать свой API каким-нибудь HTTP клиентом и посмотреть как логируются данные в консоли.

drewbal replied:
IT IS ALIVE!!! MUAHAHAHA!!!

Работает :))

RuslanZinovyev commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,96 @@
+                () -> new EntityNotFoundException(String.format("Vacancy not found by id: %s", vacancyId)));
+    }
+
+    private Vacancy toEntityFromDto(VacancyDto dto) {
```
Имя mapToEntity более подходит так как в vacancyMapper у тебя похожий метод называется просто toEntity() старайся придерживаться какой то одной стратегии в именовании переменных, методов и классов. Помни, самое сложное в программировании - это именовать переменные =)

RuslanZinovyev commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,96 @@
+        return vacancy;
+    }
+
+    private List<Long> getRejectedCandidatesIds(Long vacancyId) {
```
Ты уверен что данный метод делает то, что декларирует?
Я сильно не погружался, но не должен ли он делать что-то вроде такой логики? 
```
private List<Long> getRejectedCandidatesIds(Long vacancyId) {
    return getCandidatesByVacancyId(vacancyId).stream()
            .filter(candidate -> candidate.getCandidateStatus() == CandidateStatus.REJECTED)
            .map(Candidate::getId)
            .toList();
}
```

drewbal replied:
Согласен. Ломал голову над названием, можно изменить. Дело в том, что у кандидатов 3 статуса: APPROVED, REJECTED и PENDING. Но при закрытии вакансии APPROVED у нас должны стать TeamMembers и прекратить соискаться, а значит можно просто взять и удалить всех соискателей по этой вакансии.

drewbal replied:
И как мне кажется этот метод тогда вообще не нужен. Раз уже есть один достать всех кандидатов по вакансии

RuslanZinovyev commented on src/test/java/faang/school/projectservice/service/VacancyServiceTest.java:
```diff
@@ -0,0 +1,288 @@
+        assertThrows(EntityNotFoundException.class, () -> vacancyService.create(dto));
+
+        verify(projectValidator, times(1)).validateProjectExistsById(dto.getProjectId());
+        verify(vacancyValidator, times(1)).validateVacancyCreatorRole(dto);
```
Ты также можешь проверить что метод save() не вызывается при таком сценарии

drewbal replied:
Yes!

RuslanZinovyev commented on src/main/java/faang/school/projectservice/validator/VacancyValidator.java:
```diff
@@ -0,0 +1,50 @@
+
+    public void validateCandidateCountForClosure(Vacancy vacancy) {
+        if (vacancy.getCandidates().size() < vacancy.getCount()) {
+            throw new InsufficientCandidatesException(
```
Я вижу что ты выбрасываешь красивые кастомные исключения и это замечательно, но кто их собственно обрабаывает, где exception handlers? 
Вот допустим пользователь твоего API сделал ошибку и система выбросила исключение, назад он получит уродливое сообщение вместо нормального HTTP ответа с подробным описанием проблемы, которую он мог бы исправить. А как разработчики поймут что пошло не так, если условный стэк трейс нигде не залогирован?
Напиши мне красивый Global Exception Handling with @ControllerAdvice

YTXIRE replied:
Global Exception Handling уже есть в мастере

drewbal replied:
Сделал :) Спасибо, это круто!!

RuslanZinovyev replied:
Если Global Exception Handler уже существует то новый писать не надо, но убедись что он также покрывает твои исключения.

YTXIRE commented on src/main/java/faang/school/projectservice/dto/vacancy/FilterVacancyDto.java:
```diff
@@ -0,0 +1,23 @@
+@Builder
+public class FilterVacancyDto {
+
+    @NotBlank
```
Не хвататет сообщений валидации в данной дто

drewbal replied:
Добавил, спасибо

YTXIRE commented on src/main/java/faang/school/projectservice/exception/InsufficientCandidatesException.java:
```diff
@@ -0,0 +1,7 @@
+package faang.school.projectservice.exception;
+
+public class InsufficientCandidatesException extends RuntimeException {
```
Подлей мастер и зарегистрируй исключение в глобальном

drewbal replied:
Йес!

YTXIRE commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,96 @@
+    private final ProjectValidator projectValidator;
+    private final List<Filter<Vacancy, FilterVacancyDto>> vacancyFilters;
+
+    @Transactional
```
А зачем тебе тут @Transactional, у тебя же одна операция по сохранению?

drewbal replied:
Согласен, снял

YTXIRE commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,96 @@
+        return vacancyMapper.toDto(vacancy);
+    }
+
+    @Transactional
```
Тут такой же вопрос, она не сильно нужна тут

drewbal replied:
Тут уже нужнее, делаем удаления из разных баз. Без Transactional может произойти конфуз

RuslanZinovyev replied:
Я согласен с Андреем, у нас несколько запросов  в базу и лучше сделать этот вызов атомарным

YTXIRE commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,96 @@
+        Stream<Vacancy> vacancies = vacancyRepository.findAll().stream();
+        return vacancyFilters.stream()
+                .filter(filter -> filter.isApplicable(filters))
+                .flatMap(filter -> filter.apply(vacancies, filters))
```
У тебя есть вложенные листы, что ты используешь flatMap?

drewbal replied:
Он очень удобен. Смотри, у нас поток стримов, в функция apply на выходе тоже дает стримы, и получается Stream<Stream<Object>>. А flatMap позволяет внутренние стримы объединить, получая Stream<Object>, как нам и надо

drewbal committed:
BJS2-40651 Feature: Filter Vacancy: review fixes

VacancyService: incapsulation improved by making some methods private
FilterVacancyDTO: validation annotation refactored
ProjectService: unneccessary method removed

drewbal committed:
BJS2-40651 Feature: Create Vacancy: review fixes

VacancyController: logging added
ServiceController: logging added
GlobalExceptionHandler: fixed wrong import, logging added
Other: spaces fixed

drewbal committed:
Merge branch 'BJS2-40651-create-vacancy' into BJS2-40651-update-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/service/VacancyService.java

drewbal committed:
BJS2-40651 Feature: Create Vacancy: review fixes

VacancyController: logging added
ServiceController: logging added
GlobalExceptionHandler: InsufficientCandidates exception added
Other: Exception imports fixed

drewbal committed:
Merge branch 'BJS2-40651-update-vacancy' into BJS2-40651-delete-vacancy

drewbal committed:
BJS2-40651 Feature: Delete Vacancy: review fixes

VacancyController: logging added

drewbal committed:
Merge branch 'BJS2-40651-delete-vacancy' into BJS2-40651-filter-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/service/VacancyService.java

drewbal committed:
BJS2-40651 Feature: Filter Vacancy: review fixes

VacancyController: logging added
VacancyService: @Transactional from create method removed

drewbal committed:
Merge branch 'BJS2-40651-filter-vacancy' into BJS2-40651-get-vacancy

# Conflicts:
#	src/test/java/faang/school/projectservice/service/VacancyServiceTest.java

drewbal committed:
BJS2-40651 Feature: Get Vacancy: review fixes

VacancyController: logging added
VacancyService: deleteVacancy method logic changed: getRejectedCandidates method removed as a double of getCandidates method

drewbal committed:
BJS2-40651 Feature: Create Vacancy: optimization

NewVacancyDto: new DTO for vacancy creation
VacancyController: createVacancy method accepts new dto with minimum required fields
VacancyMapper: takes new vacancy DTO, returns default DTO
VacancyService: create method optimized for new dto
VacancyValidator: role validation methods united

drewbal committed:
Merge branch 'BJS2-40651-create-vacancy' into BJS2-40651-update-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/controller/VacancyController.java
#	src/main/java/faang/school/projectservice/mapper/VacancyMapper.java
#	src/main/java/faang/school/projectservice/service/VacancyService.java

drewbal committed:
BJS2-40651 Feature: Update Vacancy: optimization

VacancyUpdateDto: new DTO for vacancy update
VacancyController: updateVacancy method accepts new dto with minimum required fields
VacancyService: update method optimized for new dto
VacancyValidator: obsolete role validation method removed

drewbal committed:
Merge branch 'BJS2-40651-update-vacancy' into BJS2-40651-delete-vacancy

# Conflicts:
#	src/test/java/faang/school/projectservice/service/VacancyServiceTest.java

drewbal committed:
BJS2-40651 Feature: Delete Vacancy: minor fixes

Tests fixes

drewbal committed:
Merge branch 'BJS2-40651-delete-vacancy' into BJS2-40651-filter-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/controller/VacancyController.java
#	src/main/java/faang/school/projectservice/service/VacancyService.java
#	src/test/java/faang/school/projectservice/controller/VacancyControllerTest.java
#	src/test/java/faang/school/projectservice/service/VacancyServiceTest.java

drewbal committed:
BJS2-40651 Feature: Delete Vacancy: minor fixes

Tests fixes

drewbal committed:
Merge branch 'BJS2-40651-filter-vacancy' into BJS2-40651-get-vacancy

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/vacancy/NewVacancyDto.java:
```diff
@@ -0,0 +1,48 @@
+    @Size(min = 10, max = 600, message = "Vacancy description must be at least 10 and under 600 characters long")
+    private String description;
+
+    @NotNull(message = "Project id must be a positive integer")
```
Сообщение очевидно должно быть другим.

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/vacancy/NewVacancyDto.java:
```diff
@@ -0,0 +1,48 @@
+
+    @NotNull(message = "Created by id cannot be empty")
+    @Positive(message = "Created by id must be a positive integer")
+    private Long createdBy;
```
createdBy довольно странное имя переменной для типа Long. Может лучше тогда createdById?

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/vacancy/NewVacancyDto.java:
```diff
@@ -0,0 +1,48 @@
+
+    @NotNull(message = "Slavery is prohibited")
+    @Positive(message = "Salary must be a positive number")
+    private Double salary;
```
Я думаю что на данном этапе это не так критично, но на будущее когда дело касается финансов всегда используй BigDecimal

drewbal replied:
Согласен, но пока оставим Double, не знаю как это завязано во всем проекте, но можно будет улучшить

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/vacancy/NewVacancyDto.java:
```diff
@@ -0,0 +1,48 @@
+
+    @NotNull(message = "Required skills ids list cannot be empty")
+    @Size(min = 1, message = "Required skills ids list cannot be empty")
+    private List<Long> requiredSkillIds;
```
Можно 2 аннотации заменить на одну, она проверяет и на NULL и на пустой лист.

@NotEmpty(message = "Required skills ids list cannot be null or empty")

The NotEmpty annotation makes use of the NotNull class, and also checks that the size/length of the supplied object is greater than zero.

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/vacancy/VacancyDto.java:
```diff
@@ -0,0 +1,58 @@
+
+    @NotNull(message = "Required skills ids list cannot be empty")
+    @Size(min = 1, message = "Required skills ids list cannot be empty")
+    private List<Long> requiredSkillIds;
```
Можно 2 аннотации заменить на одну, она проверяет и на NULL и на пустой лист.

@NotEmpty(message = "Required skills ids list cannot be null or empty")

The NotEmpty annotation makes use of the NotNull class, and also checks that the size/length of the supplied object is greater than zero.

drewbal replied:
Переделал концепт этого дто под формат только для отправки VacancyResponseDto

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/vacancy/VacancyDto.java:
```diff
@@ -0,0 +1,58 @@
+
+    @NotNull(message = "Project id must be a positive integer")
+    @Positive(message = "Project id must be a positive integer")
+    private long projectId;
```
Используешь примитив, хотя раннее использовал объект Long. Нужно придерживаться одной стратегии.
Потом очевидно ты по ошибке скопировал одинакое сообщение для обеих аннотаций.

drewbal replied:
Переделал концепт этого дто под формат только для отправки VacancyResponseDto

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/vacancy/VacancyUpdateDto.java:
```diff
@@ -0,0 +1,24 @@
+@Validated
+public class VacancyUpdateDto {
+    @NotNull(message = "Vacancy id must be a positive integer")
+    @Positive(message = "Vacancy id must be a positive integer")
```
Одинаковые сообщения валидации для разных сценариев

RuslanZinovyev commented on src/main/java/faang/school/projectservice/service/CandidateService.java:
```diff
@@ -0,0 +1,15 @@
+@Service
+@RequiredArgsConstructor
+public class CandidateService {
+    private final CandidateRepository candidateRepository;
```
Сюда нужно какую-никакую валидацию с логером прикрутить.

RuslanZinovyev commented on src/main/java/faang/school/projectservice/service/ProjectService.java:
```diff
@@ -0,0 +1,16 @@
+@RequiredArgsConstructor
+public class ProjectService {
+    private final ProjectRepository projectRepository;
+
```
Тоже самое, валидация и логер.

drewbal replied:
А вот тут не очень, потому что метод репозитория getProjectById внутри себя делает проверку и кидает исключение. Валидация будет дублировать эту функциональность

RuslanZinovyev replied:
Логгер нужен не только для валидации но и для workflow, чтобы потом понимать что происходит кто куда ходил и тд, другое дело что надо следить за уровнями логирования чтобы не делать логи шумными.

RuslanZinovyev commented on src/main/java/faang/school/projectservice/dto/vacancy/VacancyUpdateDto.java:
```diff
@@ -0,0 +1,24 @@
+
+    @NotNull(message = "Updated by id cannot be empty")
+    @Positive(message = "Updated by id must be a positive integer")
+    private Long updatedBy;
```
`updatedById` более подходящее название для переменной

RuslanZinovyev commented on src/main/java/faang/school/projectservice/service/VacancyService.java:
```diff
@@ -0,0 +1,95 @@
+        vacancyValidator.validateVacancyExistsById(vacancyId);
+        getCandidatesByVacancyId(vacancyId).forEach(candidate ->
+                candidateService.deleteCandidateById(candidate.getId()));
+        vacancyRepository.deleteById(vacancyId);
```
Здесь тоже можно использовать логгер.

RuslanZinovyev commented on src/test/java/faang/school/projectservice/controller/VacancyControllerTest.java:
```diff
@@ -0,0 +1,165 @@
+    @Test
+    @DisplayName("Create a new vacancy successfully")
+    void testCreateVacancySuccess() {
+        when(vacancyService.create(newDto)).thenReturn(dto);
```
Неплохо было бы также проверить и негативные сценарии, что будет если метод выбросит исключение и как поведет себя твое приложение в этом случае. Что будет если DTO не будет содержать верных параметров, можно и валидацию потестить.

drewbal committed:
BJS2-40651 Feature: Create Vacancy: review fixes

Vacancy: createdBy -> createdById, updatedBy ->updatedById
NewVacancyDto: validation annotations fixed
VacancyDto: renamed in VacancyResponseDto to improve logic
Other: exception messages fixed

drewbal committed:
Merge branch 'BJS2-40651-create-vacancy' into BJS2-40651-update-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/controller/VacancyController.java
#	src/main/java/faang/school/projectservice/dto/vacancy/VacancyDto.java
#	src/main/java/faang/school/projectservice/service/VacancyService.java
#	src/test/java/faang/school/projectservice/controller/VacancyControllerTest.java
#	src/test/java/faang/school/projectservice/service/VacancyServiceTest.java
#	src/test/java/faang/school/projectservice/validator/VacancyValidatorTest.java

drewbal committed:
BJS2-40651 Feature: Update Vacancy: review fixes

Fixes from merge of Create Vacancy Feature

drewbal committed:
Merge branch 'BJS2-40651-update-vacancy' into BJS2-40651-delete-vacancy

# Conflicts:
#	src/test/java/faang/school/projectservice/controller/VacancyControllerTest.java
#	src/test/java/faang/school/projectservice/service/VacancyServiceTest.java

drewbal committed:
BJS2-40651 Feature: Delete Vacancy: review fixes

Fixes from merge of Update Vacancy Feature

drewbal committed:
Merge branch 'BJS2-40651-delete-vacancy' into BJS2-40651-filter-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/controller/VacancyController.java
#	src/main/java/faang/school/projectservice/service/VacancyService.java
#	src/test/java/faang/school/projectservice/controller/VacancyControllerTest.java
#	src/test/java/faang/school/projectservice/service/ProjectServiceTest.java
#	src/test/java/faang/school/projectservice/service/VacancyServiceTest.java

drewbal committed:
BJS2-40651 Feature: Filter Vacancy: review fixes

Fixes from merge of Delete Vacancy Feature

drewbal committed:
Merge branch 'BJS2-40651-filter-vacancy' into BJS2-40651-get-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/service/VacancyService.java
#	src/test/java/faang/school/projectservice/controller/VacancyControllerTest.java
#	src/test/java/faang/school/projectservice/service/VacancyServiceTest.java

drewbal committed:
BJS2-40651 Feature: Get Vacancy: review fixes

Review fixes + fixes from merge of Filter Vacancy Feature

drewbal committed:
Merge branch 'Dragon-master-stream7' into BJS2-40651-get-vacancy

# Conflicts:
#	src/main/java/faang/school/projectservice/service/ProjectService.java
#	src/test/java/faang/school/projectservice/service/ProjectServiceTest.java

drewbal committed:
BJS2-40651 Feat: Vacancy : conflict fixes
